packages/react/src/components/FocusTrapZone/FocusTrapZone.tsx

```diff
@@ -97,284 +92,230 @@ export const FocusTrapZone: React.FunctionComponent<IFocusTrapZoneProps> & {
-        // on the event
-        relatedTarget = doc!.activeElement as Element;
-      }
-      if (!elementContains(root.current, relatedTarget as HTMLElement)) {
-        internalState.hasFocus = false;
-      }
-    },
-    [doc, internalState, onBlurCapture],
-  );
+  /** Root div focus handler (doesn't need useCallback since it's for a native element) */
+  const onRootFocusCapture = (ev: React.FocusEvent<HTMLDivElement>) => {
+    props.onFocusCapture?.(ev);
 
-  const onRootFocusCapture = React.useCallback(
-    (ev: React.FocusEvent<HTMLDivElement>) => {
-      onFocusCapture?.(ev);
+    if (ev.target === firstBumper.current) {
+      focusBumper(true);
+    } else if (ev.target === lastBumper.current) {
+      focusBumper(false);
+    }
 
-      if (ev.target === firstBumper.current) {
-        onBumperFocus(true);
-      } else if (ev.target === lastBumper.current) {
-        onBumperFocus(false);
-      }
+    internalState.hasFocus = true;
 
-      internalState.hasFocus = true;
+    if (ev.target !== ev.currentTarget && !(ev.target === firstBumper.current || ev.target === lastBumper.current)) {
+      // every time focus changes within the trap zone, remember the focused element so that
+      // it can be restored if focus leaves the pane and returns via keystroke (i.e. via a call to this.focus(true))
+      internalState.previouslyFocusedElementInTrapZone = ev.target as HTMLElement;
+    }
+  };
 
-      if (ev.target !== ev.currentTarget && !(ev.target === firstBumper.current || ev.target === lastBumper.current)) {
-        // every time focus changes within the trap zone, remember the focused element so that
-        // it can be restored if focus leaves the pane and returns via keystroke (i.e. via a call to this.focus(true))
-        internalState.previouslyFocusedElementInTrapZone = ev.target as HTMLElement;
-      }
-    },
-    [onFocusCapture, internalState, onBumperFocus],
-  );
+  /** Called to restore focus on unmount or props change. (useEventCallback ensures latest prop values are used.) */
+  const returnFocusToInitiator = useEventCallback((elementToFocusOnDismiss: HTMLElement | null) => {
+    FocusTrapZone.focusStack = FocusTrapZone.focusStack!.filter(value => internalState.focusStackId !== value);
 
-  const returnFocusToInitiator = React.useCallback((): void => {
-    FocusTrapZone.focusStack = FocusTrapZone.focusStack.filter((value: any) => {
-      return id !== value;
-    });
-
-    if (doc) {
-      const activeElement = doc.activeElement as HTMLElement;
-      if (
-        !ignoreExternalFocusing &&
-        internalState.previouslyFocusedElementOutsideTrapZone &&
-        typeof internalState.previouslyFocusedElementOutsideTrapZone.focus === 'function' &&
-        (elementContains(root.current, activeElement) || activeElement === doc.body)
-      ) {
-        if (
-          !(
-            internalState.previouslyFocusedElementOutsideTrapZone === firstBumper.current ||
```

Comment:
This check was moved back into focusElementAsync (to ensure it's done everywhere it's needed)